README.md
[![Build Status](https://travis-ci.org/akiym/Smart-Args-TypeTiny.svg?branch=master)](https://travis-ci.org/akiym/Smart-Args-TypeTiny)
# NAME

Smart::Args::TypeTiny - We are smart, smart for you

# SYNOPSIS

    use Smart::Args::TypeTiny;

    sub func2 {
        args my $p => 'Int',
             my $q => {isa => 'Int', optional => 1},
             ;
    }
    func2(p => 3, q => 4); # p => 3, q => 4
    func2(p => 3);         # p => 3, q => undef

    sub func3 {
        args my $p => {isa => 'Int', default => 3};
    }
    func3(p => 4); # p => 4
    func3();       # p => 3

    package F;
    use Moo;
    use Smart::Args::TypeTiny;
    use Types::Standard -all;

    sub method {
        args my $self,
             my $p => Int,
             ;
    }
    sub class_method {
        args my $class => ClassName,
             my $p     => Int,
             ;
    }

    sub simple_method {
        args_pos my $self, my $p;
    }

    my $f = F->new();
    $f->method(p => 3);

    F->class_method(p => 3);

    F->simple_method(3);

# DESCRIPTION

Smart::Args::TypeTiny provides [Smart::Args](https://metacpan.org/pod/Smart::Args)-like argument validator using [Type::Tiny](https://metacpan.org/pod/Type::Tiny).

## IMCOMPATIBLE CHANGES WITH Smart::Args

- Unexpected parameters will be a fatal error

        sub foo {
            args my $x => 'Str';
        }

        sub bar {
            args_pos my $x => 'Str';
        }

        foo(x => 'a', y => 'b'); # fatal: Unexpected parameter 'y' passed
        bar('a', 'b');           # fatal: Too many parameters passed

- Optional allows to pass undef to parameter

        sub foo {
            args my $p => 'Int';
        }

        foo();           # $p = undef
        foo(p => 1);     # $p = 1
        foo(p => undef); # $p = undef

- Default parameter can take coderef as lazy value

        sub foo {
            args my $p => {isa => 'Foo', default => create_foo},         # calls every time even if $p is passed
                 my $q => {isa => 'Foo', default => sub { create_foo }}, # calls only when $p is not passed
                 ;
        }

# FUNCTIONS

- my $args = args my $var\[, $rule\], ...;

        sub foo {
            args my $int   => 'Int',
                 my $foo   => 'Foo',
                 my $bar   => {isa => 'Bar',  default  => sub { Bar->new }},
                 my $baz   => {isa => 'Baz',  optional => 1},
                 my $bool  => {isa => 'Bool', default  => 0},
                 ;

            ...
        }

        foo(int => 1, foo => Foo->new, bool => 1);

    Check parameters and fills them into lexical variables. All the parameters are mandatory by default.
    The hashref of all parameters is returned.

    `$rule` can be any one of type name, [Type::Tiny](https://metacpan.org/pod/Type::Tiny)'s type constraint object, or hashref have these parameters:

    - isa `Str|Object`

        Type name or [Type::Tiny](https://metacpan.org/pod/Type::Tiny)'s type constraint.

        - Types::Standard

                args my $int => {isa => 'Int'};

        - Mouse::Util::TypeConstraints

            Enable if Mouse.pm is loaded.

                use Mouse::Util::TypeConstraints;

                subtype 'PositiveInt',
                    as 'Int',
                    where { $_ > 0 },
                    message { 'Must be greater than zero' };

                args my $positive_int => {isa => 'PositiveInt'};

        - class name

                {
                    package Foo;
                    ...
                }

                args my $foo => {isa => 'Foo'};

    - does `Str|Object`

        Role name or [Type::Tiny](https://metacpan.org/pod/Type::Tiny)'s type constraint object.

    - optional `Bool`

        The parameter doesn't need to be passed when [optional](https://metacpan.org/pod/optional) is true.

    - default `Any|CodeRef`

        Default value for the parameter.

- my $args = args\_pos my $var\[, $rule\], ...;

        sub bar {
            args_pos my $x => 'Str',
                     my $p => 'Int',
                     ;

            ...
        }

        bar('abc', 123);

    Same as `args` except take arguments instead of parameters.

# TIPS

## SKIP TYPE CHECK

For optimization calling subroutine in runtime type check, you can overwrite `check_rule` like following code:

    {
        no warnings 'redefine';
        sub Smart::Args::TypeTiny::check_rule {
            my ($rule, $value, $exists, $name) = @_;
            return $value;
        }
    }

# SEE ALSO

[Smart::Args](https://metacpan.org/pod/Smart::Args), [Data::Validator](https://metacpan.org/pod/Data::Validator), [Params::Validate](https://metacpan.org/pod/Params::Validate), [Params::ValidationCompiler](https://metacpan.org/pod/Params::ValidationCompiler)

[Type::Tiny](https://metacpan.org/pod/Type::Tiny), [Types::Standard](https://metacpan.org/pod/Types::Standard), [Mouse::Util::TypeConstraints](https://metacpan.org/pod/Mouse::Util::TypeConstraints)

# LICENSE

Copyright (C) Takumi Akiyama.

This library is free software; you can redistribute it and/or modify
it under the same terms as Perl itself.

# AUTHOR

Takumi Akiyama <t.akiym@gmail.com>


Changes
Revision history for Perl extension Smart-Args-TypeTiny

{{$NEXT}}

0.14 2023-12-01T12:49:54Z

    - add no_check_rule function (#5 @kfly8++)

0.13 2022-08-16T19:28:32Z

    - fix failing test with Type::Tiny >= 1.016005

0.12 2017-10-29T13:41:32Z

    - stable Smart::Args::TypeTiny::Check API
    - depend Test2::V0 instead of Test2::Suite

0.11 2017-10-26T15:19:33Z

    - require Type::Tiny >= 1.002001
    - optimize checking rule parameter
    - copy hashref to prevent change original hashref

0.10 2017-10-26T04:33:54Z

    - isa/does can take type constraint string
    - support Mouse::Util::TypeConstraints for Smart::Args compatible
    - args/args_pos return hashref of parameters
    - optional parameter allows to pass undef
    - coderef of default parameter returns as scalar context

0.02 2017-09-29T04:29:56Z

    - require Test2::Suite >= 0.000077 for Test2::V0

0.01 2017-09-28T05:18:37Z

    - original version



lib/Smart/Args/TypeTiny/Check.pm
package Smart::Args::TypeTiny::Check;
use strict;
use warnings;
use Carp ();
use Scalar::Util qw/blessed/;
use Type::Registry;
use Type::Utils;

use Exporter 'import';
our @EXPORT_OK = qw/check_rule check_type/;

$Carp::CarpInternal{+__PACKAGE__}++;

my $reg = Type::Registry->for_class(__PACKAGE__);

sub check_rule {
    my ($rule, $value, $exists, $name) = @_;

    $rule = parameter_rule($rule, $name);

    my $type = rule_to_type($rule);
    if ($exists) {
        return $value if !defined $value && $rule->{optional};
        return check_type($type, $value, $name);
    } else {
        if (exists $rule->{default}) {
            my $default = $rule->{default};
            return check_type($type, ref $default eq 'CODE' ? scalar $default->() : $default, $name);
        } elsif (!$rule->{optional}) {
            Carp::confess("Required parameter '$name' not passed");
        }
    }
    return $value;
}

sub check_type {
    my ($type, $value, $name) = @_;
    return $value unless $type;
    return $value if $type->check($value);

    if ($type->has_coercion) {
        $value = $type->coerce($value);
        if ($type->check($value)) {
            return $value;
        }
    }

    Carp::confess("Type check failed in binding to parameter '\$$name'; " . $type->get_message($value));
}

sub parameter_rule {
    my ($rule, $name) = @_;

    return {isa => $rule} unless ref $rule eq 'HASH';

    my %check = map { ($_ => undef) } keys %$rule;
    delete $check{$_} for qw/isa does optional default/;

    if (%check) {
        Carp::croak("Malformed rule for '$name' (isa, does, optional, default)");
    }

    return $rule;
}

sub rule_to_type {
    my ($rule) = @_;

    if (exists $rule->{isa}) {
        my $isa = $rule->{isa};
        return $isa if blessed($isa);
        if (my $type = $reg->simple_lookup($isa)) {
            return $type;
        } else {
            my $type = Type::Utils::dwim_type(
                $isa,
                fallback => ['lookup_via_mouse', 'make_class_type'],
            );
            $type->{display_name} = $isa;
            $reg->add_type($type, $isa);
            return $type;
        }
    } elsif (exists $rule->{does}) {
        my $does = $rule->{does};
        return $does if blessed($does);
        if (my $type = $reg->simple_lookup($does)) {
            return $type;
        } else {
            my $type = Type::Utils::dwim_type(
                $does,
                fallback => ['make_role_type'],
            );
            $type->{display_name} = $does;
            $reg->add_type($type, $does);
            return $type;
        }
    }
    return undef;
}

1;


lib/Smart/Args/TypeTiny.pm
package Smart::Args::TypeTiny;
use strict;
use warnings;
our $VERSION = "0.13";
use Carp ();
use PadWalker qw/var_name/;

use Exporter 'import';
our @EXPORT = qw/args args_pos/;

$Carp::CarpInternal{+__PACKAGE__}++;

use Smart::Args::TypeTiny::Check qw/check_rule/;

my %is_invocant = map { ($_ => 1) } qw($self $class);

sub args {
    {
        package DB;
        # call of caller in DB package sets @DB::args,
        # which requires list context, but we don't need return values
        () = CORE::caller(1);
    }

    if (@_) {
        my $name = var_name(1, \$_[0]) || '';
        if ($is_invocant{$name}) { # seems instance/class method call
            $name =~ s/^\$//;
            $_[0] = shift @DB::args;
            if (defined $_[1]) { # has rule?
                $_[0] = check_rule($_[1], $_[0], 1, $name);
                shift;
            }
            shift;
        }
    }

    my $args = (@DB::args == 1 && ref $DB::args[0] eq 'HASH')
            ? +{ %{$DB::args[0]} } # must be hash
            : +{ @DB::args };      # must be key-value list
    my $kv = {};

    # args my $var => RULE
    #         ~~~~    ~~~~
    #         undef   defined

    for (my $i = 0; $i < @_; $i++) {
        (my $name = var_name(1, \$_[$i]))
            or Carp::croak('Usage: args my $var => RULE, ...');
        $name =~ s/^\$//;

        # with rule (my $foo => RULE, ...)
        if (defined $_[$i+1]) {
            $_[$i] = $kv->{$name} = check_rule($_[$i+1], $args->{$name}, exists $args->{$name}, $name);
            delete $args->{$name};
            $i++;
        }
        # without rule (my $foo, my $bar, ...)
        else {
            unless (exists $args->{$name}) {
                Carp::confess("Required parameter '$name' not passed");
            }
            $_[$i] = $kv->{$name} = delete $args->{$name};
        }
    }

    for my $name (sort keys %$args) {
        Carp::confess("Unexpected parameter '$name' passed");
    }

    return $kv;
}

sub args_pos {
    {
        package DB;
        # call of caller in DB package sets @DB::args,
        # which requires list context, but we don't need return values
        () = CORE::caller(1);
    }

    if (@_) {
        my $name = var_name(1, \$_[0]) || '';
        if ($is_invocant{$name}) { # seems instance/class method call
            $name =~ s/^\$//;
            $_[0] = shift @DB::args;
            if (defined $_[1]) { # has rule?
                $_[0] = check_rule($_[1], $_[0], 1, $name);
                shift;
            }
            shift;
        }
    }

    my $args = [@DB::args];
    my $kv = {};

    # args my $var => RULE
    #         ~~~~    ~~~~
    #         undef   defined

    for (my $i = 0; $i < @_; $i++) {
        (my $name = var_name(1, \$_[$i]))
            or Carp::croak('Usage: args_pos my $var => RULE, ...');
        $name =~ s/^\$//;

        # with rule (my $foo => RULE, ...)
        if (defined $_[$i+1]) {
            $_[$i] = $kv->{$name} = check_rule($_[$i+1], $args->[0], @$args > 0, $name);
            shift @$args;
            $i++;
        }
        # without rule (my $foo, my $bar, ...)
        else {
            unless (@$args > 0) {
                Carp::confess("Required parameter '$name' not passed");
            }
            $_[$i] = $kv->{$name} = shift @$args;
        }
    }

    if (@$args) {
        Carp::confess('Too many parameters passed');
    }

    return $kv;
}

1;
__END__

=encoding utf-8

=head1 NAME

Smart::Args::TypeTiny - We are smart, smart for you

=head1 SYNOPSIS

    use Smart::Args::TypeTiny;

    sub func2 {
        args my $p => 'Int',
             my $q => {isa => 'Int', optional => 1},
             ;
    }
    func2(p => 3, q => 4); # p => 3, q => 4
    func2(p => 3);         # p => 3, q => undef

    sub func3 {
        args my $p => {isa => 'Int', default => 3};
    }
    func3(p => 4); # p => 4
    func3();       # p => 3

    package F;
    use Moo;
    use Smart::Args::TypeTiny;
    use Types::Standard -all;

    sub method {
        args my $self,
             my $p => Int,
             ;
    }
    sub class_method {
        args my $class => ClassName,
             my $p     => Int,
             ;
    }

    sub simple_method {
        args_pos my $self, my $p;
    }

    my $f = F->new();
    $f->method(p => 3);

    F->class_method(p => 3);

    F->simple_method(3);

=head1 DESCRIPTION

Smart::Args::TypeTiny provides L<Smart::Args>-like argument validator using L<Type::Tiny>.

=head2 IMCOMPATIBLE CHANGES WITH Smart::Args

=over 4

=item Unexpected parameters will be a fatal error

    sub foo {
        args my $x => 'Str';
    }

    sub bar {
        args_pos my $x => 'Str';
    }

    foo(x => 'a', y => 'b'); # fatal: Unexpected parameter 'y' passed
    bar('a', 'b');           # fatal: Too many parameters passed

=item Optional allows to pass undef to parameter

    sub foo {
        args my $p => {isa => 'Int', optional => 1};
    }

    foo();           # $p = undef
    foo(p => 1);     # $p = 1
    foo(p => undef); # $p = undef

=item Default parameter can take coderef as lazy value

    sub foo {
        args my $p => {isa => 'Foo', default => create_foo},         # calls every time even if $p is passed
             my $q => {isa => 'Foo', default => sub { create_foo }}, # calls only when $p is not passed
             ;
    }

=back

=head1 FUNCTIONS

=over 4

=item my $args = args my $var[, $rule], ...;

    sub foo {
        args my $int   => 'Int',
             my $foo   => 'Foo',
             my $bar   => {isa => 'Bar',  default  => sub { Bar->new }},
             my $baz   => {isa => 'Baz',  optional => 1},
             my $bool  => {isa => 'Bool', default  => 0},
             ;

        ...
    }

    foo(int => 1, foo => Foo->new, bool => 1);

Check parameters and fills them into lexical variables. All the parameters are mandatory by default.
The hashref of all parameters is returned.

C<$rule> can be any one of type name, L<Type::Tiny>'s type constraint object, or hashref have these parameters:

=over 4

=item isa C<Str|Object>

Type name or L<Type::Tiny>'s type constraint.

=over 4

=item Types::Standard

    args my $int => {isa => 'Int'};

=item Mouse::Util::TypeConstraints

Enable if Mouse.pm is loaded.

    use Mouse::Util::TypeConstraints;

    subtype 'PositiveInt',
        as 'Int',
        where { $_ > 0 },
        message { 'Must be greater than zero' };

    args my $positive_int => {isa => 'PositiveInt'};

=item class name

    {
        package Foo;
        ...
    }

    args my $foo => {isa => 'Foo'};

=back

=item does C<Str|Object>

Role name or L<Type::Tiny>'s type constraint object.

=item optional C<Bool>

The parameter doesn't need to be passed when L<optional> is true.

=item default C<Any|CodeRef>

Default value for the parameter.

=back

=item my $args = args_pos my $var[, $rule], ...;

    sub bar {
        args_pos my $x => 'Str',
                 my $p => 'Int',
                 ;

        ...
    }

    bar('abc', 123);

Same as C<args> except take arguments instead of parameters.

=back

=head1 TIPS

=head2 SKIP TYPE CHECK

For optimization calling subroutine in runtime type check, you can overwrite C<check_rule> like following code:

    {
        no warnings 'redefine';
        *Smart::Args::TypeTiny::check_rule = \&Smart::Args::TypeTiny::Check::no_check_rule;
    }

C<Smart::Args::TypeTiny::Check::no_check_rule> is a function without type checking and type coerce, but settings such as default and optional work the same as C<check_rule>.

=head1 SEE ALSO

L<Smart::Args>, L<Data::Validator>, L<Params::Validate>, L<Params::ValidationCompiler>

L<Type::Tiny>, L<Types::Standard>, L<Mouse::Util::TypeConstraints>

=head1 LICENSE

Copyright (C) Takumi Akiyama.

This library is free software; you can redistribute it and/or modify
it under the same terms as Perl itself.

=head1 AUTHOR

Takumi Akiyama E<lt>t.akiym@gmail.comE<gt>

=cut
